File: java/src/main/java/io/fivexer/sdk/JoinLinks.java
package io.fivexer.sdk;

import io.fivexer.sdk.internal.Json;
import io.fivexer.sdk.model.CreateJoinLink;
import io.fivexer.sdk.model.CreateJoinLinkResult;
import io.fivexer.sdk.model.JoinLink;
import io.fivexer.sdk.model.JoinLinkList;
import java.util.Collections;
import java.util.List;

/** QR join links: worker self-registration with preset tags, skills and team. */
public final class JoinLinks {

    private final Fivexer client;

    JoinLinks(Fivexer client) {
        this.client = client;
    }

    /**
     * Create a link. The {@code joinUrl} on the result is returned only here — a lost link is
     * re-created, never recovered.
     */
    public CreateJoinLinkResult create(CreateJoinLink input) {
        return client.request("POST", "/join-links", input.toJson(), null,
                CreateJoinLinkResult.class, false);
    }

    public List<JoinLink> list() {
        JoinLinkList envelope = client.request("GET", "/join-links", null, null,
                JoinLinkList.class, false);
        return envelope.getLinks() == null ? Collections.emptyList() : envelope.getLinks();
    }

    public void revoke(String linkId) {
        client.request("DELETE", "/join-links/" + Json.enc(linkId), null, null, Void.class, true);
    }
}

File: php/src/Resource/JoinLinks.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Resource;

use Fivexer\SDK\Fivexer;
use Fivexer\SDK\Internal\Json;
use Fivexer\SDK\Model\CreateJoinLink;
use Fivexer\SDK\Model\CreateJoinLinkResult;
use Fivexer\SDK\Model\JoinLink;

/** QR join links: worker self-registration with preset tags, skills and team. */
final class JoinLinks
{
    public function __construct(private readonly Fivexer $client)
    {
    }

    /**
     * Create a link. The `joinUrl` on the result is returned only here — a lost link is
     * re-created, never recovered.
     */
    public function create(CreateJoinLink $input): CreateJoinLinkResult
    {
        return CreateJoinLinkResult::fromArray(
            $this->client->request('POST', '/join-links', $input->toArray()) ?? []
        );
    }

    /** @return list<JoinLink> */
    public function list(): array
    {
        $data = $this->client->request('GET', '/join-links') ?? [];
        /** @var list<JoinLink> */
        return Json::parseEach($data, 'links', [JoinLink::class, 'fromArray']);
    }

    public function revoke(string $linkId): void
    {
        $this->client->request('DELETE', '/join-links/' . \rawurlencode($linkId));
    }
}
